lib/auth/new_web_user.go

```diff
@@ -53,69 +56,79 @@ func (s *AuthServer) CreateSignupToken(userv1 services.UserV1) (string, error) {
 			return "", trace.Wrap(err)
 		}
 	}
-	// check existing
+
+	// TODO(rjones): TOCTOU, instead try to create signup token for user and fail
+	// when unable to.
 	_, err := s.GetPasswordHash(user.GetName())
 	if err == nil {
-		return "", trace.BadParameter("user '%v' already exists", user)
+		return "", trace.BadParameter("user %q already exists", user)
 	}
 
 	token, err := utils.CryptoRandomHex(TokenLenBytes)
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
 
-	otp, err := hotp.GenerateHOTP(defaults.HOTPTokenDigits, false)
+	accountName := user.GetName() + "@" + s.AuthServiceName
+	otpKey, otpQRCode, err := s.initializeTOTP(accountName)
 	if err != nil {
-		log.Errorf("[AUTH API] failed to generate HOTP: %v", err)
 		return "", trace.Wrap(err)
 	}
-	otpQR, err := otp.QR("Teleport: " + user.GetName() + "@" + s.AuthServiceName)
-	if err != nil {
-		return "", trace.Wrap(err)
+
+	// create and upsert signup token
+	tokenData := services.SignupToken{
+		Token:     token,
+		User:      userv1,
+		OTPKey:    otpKey,
+		OTPQRCode: otpQRCode,
 	}
 
-	otpMarshalled, err := hotp.Marshal(otp)
+	err = s.UpsertSignupToken(token, tokenData, defaults.MaxSignupTokenTTL)
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
 
-	otpFirstValues := make([]string, defaults.HOTPFirstTokensRange)
-	for i := 0; i < defaults.HOTPFirstTokensRange; i++ {
-		otpFirstValues[i] = otp.OTP()
-	}
+	log.Infof("[AUTH API] created the signup token for %q", user)
+	return token, nil
+}
 
-	tokenData := services.SignupToken{
-		Token:           token,
-		User:            userv1,
-		Hotp:            otpMarshalled,
-		HotpFirstValues: otpFirstValues,
-		HotpQR:          otpQR,
+// initializeTOTP creates TOTP algorithm and returns the key and QR code.
+func (s *AuthServer) initializeTOTP(accountName string) (key string, qr []byte, err error) {
+	// create totp key
+	otpKey, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      "Teleport",
+		AccountName: accountName,
+	})
+	if err != nil {
+		return "", nil, err
```

Comment:
please add `trace.Wrap` around this error here and in similar places so we can track stack traces returned